fvm/evm/stdlib/contract.go

```diff
@@ -38,6 +40,180 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
 var evmAddressBytesStaticType = interpreter.ConvertSemaArrayTypeToStaticArrayType(nil, evmAddressBytesType)
 var EVMAddressBytesCadenceType = cadence.NewConstantSizedArrayType(types.AddressLength, cadence.TheUInt8Type)
 
+// EVM.encodeABI
+
+const internalEVMTypeEncodeABIFunctionName = "encodeABI"
+
+var internalEVMTypeEncodeABIFunctionType = &sema.FunctionType{
+	Parameters: []sema.Parameter{
+		{
+			Label:      sema.ArgumentLabelNotRequired,
+			Identifier: "abiSpec",
+			TypeAnnotation: sema.NewTypeAnnotation(
+				sema.StringType,
+			),
+		},
+		{
+			Label: "arguments",
+			TypeAnnotation: sema.NewTypeAnnotation(
+				sema.NewVariableSizedType(nil, sema.AnyStructType),
+			),
+		},
+	},
+	ReturnTypeAnnotation: sema.NewTypeAnnotation(sema.ByteArrayType),
+}
+
+func newInternalEVMTypeEncodeABIFunction(
+	gauge common.MemoryGauge,
+) *interpreter.HostFunctionValue {
+	return interpreter.NewHostFunctionValue(
+		gauge,
+		internalEVMTypeEncodeABIFunctionType,
+		func(invocation interpreter.Invocation) interpreter.Value {
+			inter := invocation.Interpreter
+			locationRange := invocation.LocationRange
+
+			// Get `abiSpec` argument
+
+			abiSpec, ok := invocation.Arguments[0].(*interpreter.StringValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			abi, err := gethABI.JSON(strings.NewReader(abiSpec.Str))
+			if err != nil {
+				panic(err)
+			}
+
+			// Get `arguments` argument
+
+			argumentsValue, ok := invocation.Arguments[1].(*interpreter.ArrayValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			arguments := make([]interface{}, 0)
+			for i := 0; i < argumentsValue.Count(); i++ {
+				arg := argumentsValue.Get(inter, locationRange, i)
+				switch value := arg.(type) {
+				case *interpreter.StringValue:
+					arguments = append(arguments, value.Str)
+				case interpreter.UInt64Value:
+					arguments = append(arguments, uint64(value))
+				}
+			}
+			packed, err := abi.Pack("details", arguments...)
```

Comment:
Now that I understand the ABI encoding/decoding more: Doesn't this hardcode the method name to "details"? Shouldn't the method name be passed to the `encodeABI` (and `decodeABI`) function, i.e. should those functions have an additional parameter?

Reply:
https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/66325132bde3f61d2c18748cfcd3e181b063ea00/contracts/test/mocks/MockGateway.sol#L224C32-L224C32

Here is a good example of how we plan to use ABI encode/decode on chain.  Mainly passing in a 'partial' abi to encode only those variables into bytes.  I would say encode packed is less important of a feature as it seems to do everything normal encode/decode does in less space.  It may be good to support in a future roadmap.  I think the functions you outlined are a good place to start.  It would allow developers to create data packages that would be portable between flow and evm as well as generate EVM function calls/signatures on flow.